docs/topics/request-response.rst

```diff
@@ -312,6 +312,143 @@ errors if needed::
+
+REQUEST_FINGERPRINTER_CLASS
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Default: ``'scrapy.utils.request.RequestFingerprinter'``
+
+The import path of a :ref:`request fingerprinter
+<custom-request-fingerprinter>`.
+
+The default request fingerprinter takes into account a canonical version
+(:func:`w3lib.url.canonicalize_url`) of
+:attr:`request.url <scrapy.http.Request.url>` and the values of
+:attr:`request.method <scrapy.http.Request.method>` and
+:attr:`request.body <scrapy.http.Request.body>`. It then generates an
+`SHA1 <https://en.wikipedia.org/wiki/SHA-1>`_ hash from their concatenation.
+
+
+.. _custom-request-fingerprinter:
+
+Writing your own request fingerprinter
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A request fingerprinter is a class that must implement the following method:
+
+.. method:: fingerprint(self, request)
+
+   Return a :class:`str` that uniquely identifies *request*.
+
+   The returned :class:`str` must be a valid filename. The hexadecimal
+   representation of a hash is recommended.
+
+   :param request: request to fingerprint
+   :type request: scrapy.http.Request
+
+Additionally, they may also implement the following methods:
+
+.. classmethod:: from_crawler(cls, crawler)
+
+   If present, this class method is called to create a request fingerprinter
+   instance from a :class:`~scrapy.crawler.Crawler` object. It must return a
+   new instance of the request fingerprinter.
+
+   *crawler* provides access to all Scrapy core components like settings and
+   signals; it is a way for the request fingerprinter to access them and hook
+   its functionality into Scrapy.
+
+   :param crawler: crawler that uses this pipeline
+   :type crawler: :class:`~scrapy.crawler.Crawler` object
+
+.. classmethod:: from_settings(cls, settings)
+
+   If present, and ``from_crawler`` is not defined, this class method is called
+   to create a request fingerprinter instance from a
+   :class:`~scrapy.settings.Settings` object. It must return a new instance of
+   the request fingerprinter.
+
+For some common use cases you can use
+:func:`~scrapy.utils.request.request_fingerprint` in your ``fingerprint``
+method implementation:
+
+.. autofunction:: scrapy.utils.request.request_fingerprint
+
+For example, to take the value of a request header named ``X-ID`` into
+account::
+
+    # my_project/settings.py
+    REQUEST_FINGERPRINTER_CLASS = 'my_project.utils.RequestFingerprinter'
+
+    # my_project/utils.py
+    from scrapy.utils.request import request_fingerprint
+
+    class RequestFingerprinter:
+
+        def fingerprint(self, request):
+            return request_fingerprint(request, include_headers=['X-ID'])
+
+You can also write your own fingerprinting logic from scratch. However, if you
+do not use :func:`~scrapy.utils.request.request_fingerprint`, make sure you use
+:class:`~weakref.WeakKeyDictionary` to cache request fingerprints. For example,
+to take into account only the URL of a request, without any prior URL
+canonicalization::
```

Comment:
I think it'd be good to add a line explaining the reason to use `weakref.WeakKeyDictionary`, as it might not seem obvious.